src/installer/__main__.py

```diff
@@ -0,0 +1,206 @@
+    # type: (Dict[str, List[RecordEntry]], Dict[str, str], Dict[str, Any]) -> None
+    for scheme, records in record_dict.items():
+        if scheme not in ("purelib", "platlib"):
+            continue
+        for record in records:
+            target_path = os.path.join(scheme_dict[scheme], record.path)
+            compileall.compile_file(target_path, **compile_args)
+
+
+def generate_bytecode(record_dict, scheme_dict, levels, stripdir=None):
+    # type: (Dict[str, List[RecordEntry]], Dict[str, str], Union[List[int], bool], Optional[str]) -> None
+    """Generate bytecode for Python files."""
+    if sys.version_info[0] == 2:
+        return _compile_records(record_dict, scheme_dict, {})
+
+    assert isinstance(levels, list)
+
+    compile_args = {}  # type: Dict[str, Any]
+    if stripdir:
+        compile_args["stripdir"] = stripdir
+
+    for level in levels:
+        compile_args["optimize"] = level
+        _compile_records(record_dict, scheme_dict, compile_args)
+
+
+def check_python_version(metadata):  # type: (Message) -> None
+    """Check if the project support the current interpreter."""
+    try:
+        import packaging.specifiers
+    except ImportError:
+        warnings.warn(
+            "'packaging' module not available, "
+            "skiping python version compatibility check"
+        )
+        return
+
+    requirement = metadata["Requires-Python"]
+    if not requirement:
+        return
+
+    versions = requirement.split(",")
+    for version in versions:
+        specifier = packaging.specifiers.Specifier(version)
+        if platform.python_version() not in specifier:
+            raise RuntimeError(
+                "Incompatible python version, needed: {}".format(version)
+            )
+
+
+def check_dependencies(metadata):  # type: (Message) -> None
+    """Check if the project dependencies are met."""
+    try:
+        import build
+    except ImportError:
+        warnings.warn("'build' module not available, skiping dependency check")
+        return
+
+    missing = {
+        unmet
+        for requirement in metadata.get_all("Requires-Dist") or []
+        for unmet_list in build.check_dependency(requirement)
+        for unmet in unmet_list
+    }
+    if missing:
+        missing_list = ", ".join(missing)
+        raise RuntimeError("Missing requirements: {}".format(missing_list))
+
+
+def main(cli_args, program=None):
+    # type: (Sequence[str], Optional[str]) -> None
+    """Process arguments and perform the install."""
+    parser = main_parser()
+    if program:
+        parser.prog = program
+    args = parser.parse_args(cli_args)
+
+    bytecode_stripdir = None  # type: Optional[str]
+
+    with installer.sources.WheelFile.open(args.wheel) as source:
+        # compability checks
+        metadata_contents = source.read_dist_info("METADATA")
+        metadata = installer.utils.parse_metadata_file(metadata_contents)
+        check_python_version(metadata)
```

Comment:
Sure.

Reply:
I don't have strong opinions one way or the other about a python version check, but FWIW when I originally read this code I assumed the dependency check had a skip option because it was assumed people might often be installing a wheel when its dependencies are not *currently* installed, but *will* be by the time it is expected to be used.

In contrast, skipping the python version check is saying that you think the wheel metadata is flat out wrong. Which I guess you're saying it might sometimes be...

Reply:
Yup. If you want a concrete example, see https://github.com/h5py/h5py/pull/1727 - CI is slow, so we had an incentive to stop testing on Python 3.6. When we did, we could either declare that it requires Python 3.7 (which would be wrong at least for a while), or declare that it still supports 3.6 - until someone one day realises that that's wrong and files an issue.

Being overly restrictive is generally easier for maintainers - someone who does a simply `pip install h5py` on Python 3.6 will automatically get an older version that was tested on 3.6. And if someone really needs a newer h5py, they can bypass the check and see if it works.

Reply:
> In contrast, skipping the python version check is saying that you think the wheel metadata is flat out wrong. Which I guess you're saying it might sometimes be...

The main thing here I think is that version checks need to be done against the target interpreter sysconfig(the one that will be used with the package being installed) and not the one actually executing `installer` and doing the installation(which can be entirely different for example when cross compiling)